programs/mpl-core/src/plugins/autograph.rs

```diff
@@ -0,0 +1,140 @@
+                    solana_program::msg!("Autograph: Rejected");
+                    return Err(MplCoreError::InvalidPluginOperation.into());
+                }
+            }
+            None => {
+                if &sig.address != authority {
+                    solana_program::msg!("Autograph: Rejected");
+                    return Err(MplCoreError::MissingSigner.into());
+                }
+            }
+        }
+    }
+
+    let new_signatures: HashSet<_> = new_autograph
+        .signatures
+        .iter()
+        .map(|sig| sig.address)
+        .collect();
+
+    if new_autograph.signatures.len() != new_signatures.len() {
+        solana_program::msg!("Autograph: Rejected");
+        return Err(MplCoreError::InvalidPluginSetting.into());
+    }
+
+    if !is_plugin_authority {
+        // only the plugin authority can remove signatures
+        for (key, _) in existing_map.iter() {
+            if !new_signatures.contains(key) {
+                solana_program::msg!("Autograph: Rejected");
+                return Err(MplCoreError::MissingSigner.into());
+            }
+        }
+    }
+
+    Ok(ValidationResult::Pass)
+}
+
+impl PluginValidation for Autograph {
+    fn validate_create(
+        &self,
+        ctx: &PluginValidationContext,
+    ) -> Result<ValidationResult, ProgramError> {
+        validate_autograph(self, None, ctx.authority_info.key, true)
+    }
+
+    fn validate_add_plugin(
+        &self,
+        ctx: &PluginValidationContext,
+    ) -> Result<ValidationResult, ProgramError> {
+        match ctx.target_plugin {
+            Some(Plugin::Autograph(_autograph)) => {
+                validate_autograph(self, None, ctx.authority_info.key, true)?;
+                Ok(ValidationResult::Approved)
+            }
+            _ => Ok(ValidationResult::Pass),
+        }
+    }
+
+    fn validate_update_plugin(
+        &self,
+        ctx: &PluginValidationContext,
+    ) -> Result<ValidationResult, ProgramError> {
+        let resolved_authorities = ctx
+            .resolved_authorities
+            .ok_or(MplCoreError::InvalidAuthority)?;
+        match ctx.target_plugin {
+            Some(Plugin::Autograph(autograph)) => {
+                validate_autograph(
+                    autograph,
+                    Some(self),
+                    ctx.authority_info.key,
+                    resolved_authorities.contains(ctx.self_authority),
+                )?;
+                solana_program::msg!("Autograph: Approved");
+                Ok(ValidationResult::Approved)
+            }
+            _ => Ok(ValidationResult::Pass),
+        }
+    }
+
+    /// Validate the revoke plugin authority lifecycle action.
+    fn validate_revoke_plugin_authority(
+        &self,
+        ctx: &PluginValidationContext,
+    ) -> Result<ValidationResult, ProgramError> {
+        if ctx.self_authority
+            == &(Authority::Address {
+                address: *ctx.authority_info.key,
+            })
```

Comment:
I think there might be a bug in our `validate_revoke_plugin_authority` calls across all plugins.  If you set the authority to the Update Authority for example, and then you sign as the Update Authority, since it is an owner-managed plugin it won't be approved by the asset, and since it is not a plain address, it will not be approved here, so the UpdateAuthority would not be able to self-revoke.

Will be the same for an authority-managed plugin set to Owner: Owner will not be able to self-revoke.

Reply:
hmm good point - https://linear.app/mplx/issue/MET-463/fix-plugin-self-revoke-for-non-address-plugin-authorities

Reply:
good point, created a ticket to track